tools/provenance-generator/pkg/parse.go

```diff
@@ -0,0 +1,169 @@
+	// Multiheaded repos may need to not make this call.
+	// The git fetch <remote> <BaseRef> call occurs regardless.
+	SkipFetchHead bool `json:"skip_fetch_head,omitempty"`
+}
+
+type Record struct {
+	Refs     Refs      `json:"refs"`
+	Commands []Command `json:"commands,omitempty"`
+	Failed   bool      `json:"failed,omitempty"`
+
+	// FinalSHA is the SHA from ultimate state of a cloned ref
+	// This is used to populate RepoCommit in started.json properly
+	FinalSHA string `json:"final_sha,omitempty"`
+}
+
+// Can't seem to import "k8s.io/test-infra/prow/pod-utils/clone" library :(
+type Command struct {
+	Command string `json:"command"`
+	Output  string `json:"output,omitempty"`
+	Error   string `json:"error,omitempty"`
+}
+
+func LoadParameters(config Config) Config {
+	// Parse Entrypoint of the prowjob to pull args, path, etc
+	entryPointOpsRaw := os.Getenv(config.EntryPointOptsVariable)
+	entryPointOpts := &entrypoint.Options{}
+	err := json.Unmarshal([]byte(entryPointOpsRaw), entryPointOpts)
+	if err != nil {
+		log.Fatalf("failed to unmarshal ENTRYPOINTS_OPTIONS env variable %v", err)
+	}
+	config.EntryPointOpts = entryPointOpts
+
+	// Sanity checks
+	if config.EntryPointOpts.Args[0] != "runner.sh" {
+		log.Fatal("this prowjob is misconfigured, expecting runner.sh to be called first")
+	}
+
+	config.Arguments = map[string]string{}
+	keys := []string{}
+
+	// Parse entrypoint flags
+	for _, elem := range config.EntryPointOpts.Args {
+		params := strings.Split(elem, "=") // expected values --something=foo
+		if len(params) == 1 {
+			config.Arguments[params[0]] = ""
+			keys = append(keys, params[0])
+		} else if len(params) == 2 {
+			config.Arguments[params[0]] = params[1]
+			keys = append(keys, params[0])
+		}
+	}
+	config.ArgumentsInsertOrder = keys
+
+	//Appoximate start time of job
+	fileInfo, err := os.Stat(config.CloneLogPath)
+	if err != nil {
+		log.Fatalf("failed to open file %v", err)
+	}
+	config.StartTime = fileInfo.ModTime()
+
+	byteValue, _ := os.ReadFile(config.CloneLogPath)
+	cloneLogs := []Record{}
+
+	err = json.Unmarshal(byteValue, &cloneLogs)
+	if err != nil {
+		log.Fatalf("failed to unmarshal clone-logs file %v", err)
+	}
+	config.CloneRecords = cloneLogs
+
+	// Lets strip out the empty refs that is inserted by prow
+	for i := len(config.CloneRecords) - 1; i >= 0; i-- {
+		cloneRecord := config.CloneRecords[i]
+		if cloneRecord.FinalSHA == "" {
+			config.CloneRecords = append(config.CloneRecords[:i],
+				config.CloneRecords[i+1:]...)
+		}
+	}
+
+	// Parse Image References
+	imageRefFile, err := os.ReadFile(config.ImageReferencePath)
+	if err != nil {
+		log.Fatalf("failed to open file %v", err)
+	}
+	lines := strings.Split(string(imageRefFile), "\n")
+	subjects := []in_toto.Subject{}
+	for _, elem := range lines {
+		name, err := reference.ParseNamed(elem)
+		if err != nil {
+			break
+		}
+		digest := strings.TrimPrefix(elem, name.Name()+"@")
+		digestComponents := strings.Split(digest, ":")
```

Comment:
How will we detect if this code gets confused. Should we make the digest part of the object returned from `reference.ParseNamed`, such that it could return an error in that case?

Reply:
Also, `strings.Cut` here!

Reply:
Ah, this is coming from `docker/distribution`...

Consider using [`google/go-containerregistry`](https://pkg.go.dev/github.com/google/go-containerregistry@v0.11.0/pkg/name#pkg-types), particularly the `Digest` object.

Reply:
I spent alot of time looking for a package that can parse docker image reference but the best one I found was the docker/distribution package. Still annoyed that it doesn't have a function that parses a image ref and returns hostname, digest, digest-algo, name, etc